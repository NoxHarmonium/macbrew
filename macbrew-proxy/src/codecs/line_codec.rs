use std::{io, str};
use tokio_util::codec::{Decoder, Encoder};

use bytes::BytesMut;

pub struct LineCodec;

impl Decoder for LineCodec {
    type Item = String;
    type Error = io::Error;

    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        let newline = src.as_ref().iter().position(|b| *b == b'\n');
        if let Some(n) = newline {
            let line = src.split_to(n + 1);
            return match str::from_utf8(line.as_ref()) {
                Ok(s) => Ok(Some(s.to_string())),
                Err(_) => Err(io::Error::new(io::ErrorKind::Other, "Invalid String")),
            };
        }
        Ok(None)
    }
}

impl Encoder<String> for LineCodec {
    type Error = io::Error;

    fn encode(&mut self, item: String, dst: &mut BytesMut) -> Result<(), Self::Error> {
        Ok(())
    }
}
