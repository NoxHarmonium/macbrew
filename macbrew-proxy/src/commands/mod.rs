pub mod command;
pub mod get_fermentation;
pub mod get_recipes;
pub mod get_sessions;
pub mod list_sessions;
pub mod list_steps;
pub mod ping;
