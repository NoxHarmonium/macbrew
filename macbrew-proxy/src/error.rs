use reqwest::Error as ReqwestError;
use serde::{ser, Serialize};
use snafu::Snafu;

// TODO: Clean these up to provide more context and better error messages
#[derive(Debug, Snafu, Serialize)]
pub enum Error {
    #[snafu(display("Could not serialize data"))]
    SerializationFailure {},

    #[snafu(display("Error calling API: {}", inner_error))]
    ApiError {
        #[serde(skip_serializing)]
        inner_error: ReqwestError,
    },

    #[snafu(display("API response failed validation: {}", message))]
    ApiResponseValidationError { message: String },

    #[snafu(display("Invalid command input: {}", message))]
    InvalidCommandInput { message: String },

    #[snafu(display("Serial IO error: {}", inner_error))]
    SerialIoError {
        #[serde(skip_serializing)]
        inner_error: std::io::Error,
    },

    #[snafu(display("Unknown error: {}", message))]
    UnknownError { message: String },

    #[snafu(display("Could not allocate serialization buffer. (Not enough RAM?)"))]
    SerializeBufferFull {},

    #[snafu(display("Could not work out length of sequence"))]
    SerializeSeqLengthUnknown {},
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

impl From<ReqwestError> for Error {
    fn from(inner_error: ReqwestError) -> Self {
        Error::ApiError {
            inner_error: inner_error,
        }
    }
}

impl From<std::io::Error> for Error {
    fn from(inner_error: std::io::Error) -> Self {
        Error::SerialIoError {
            inner_error: inner_error,
        }
    }
}

impl ser::Error for Error {
    #[cold]
    fn custom<T: std::fmt::Display>(msg: T) -> Error {
        // TODO: What should this actually be doing?
        // Hint: https://github.com/serde-rs/json/blob/master/src/error.rs#L373
        Error::UnknownError {
            message: msg.to_string(),
        }
    }
}
