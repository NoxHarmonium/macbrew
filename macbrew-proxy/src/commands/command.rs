use crate::data::data_manager::DataManager;
use crate::error::Result;
use crate::serializers::mac::serialize;
use async_trait::async_trait;
use serde::Serialize;

#[async_trait]
pub trait Command<DataManagerType: DataManager> {
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>>;
}

pub fn prepare_response<T>(rid: &str, success: bool, response_object: &T) -> Result<Vec<u8>>
where
    T: Serialize + ?Sized,
{
    let preable_bytes = serialize(&rid)?;
    let success_bytes = serialize(&success)?;
    let response_bytes = serialize(&response_object)?;
    Ok([preable_bytes, success_bytes, response_bytes].concat())
}
