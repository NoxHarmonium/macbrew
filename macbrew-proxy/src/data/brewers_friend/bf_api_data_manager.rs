use crate::data::brewers_friend::bf_data_manager::BFDataManager;
use crate::data::brewers_friend::recipes::BeerXml;
use crate::data::brewers_friend::sessions::{BFBrewSession, BFBrewSessionsResponse};
use crate::error::{Error, Result};
use async_trait::async_trait;
use reqwest;
use reqwest::header::HeaderMap;
use serde::Deserialize;

#[cfg(test)]
use mockito;

#[derive(Deserialize, Debug)]
struct Config {
    brewers_friend_api_key: String,
}

pub struct BFApiDataManager;

fn get_base_url() -> String {
    #[cfg(test)]
    return mockito::server_url();
    #[cfg(not(test))]
    return String::from("https://api.brewersfriend.com");
}

fn create_client() -> Result<reqwest::Client> {
    let mut headers = HeaderMap::new();

    let config = match envy::from_env::<Config>() {
        Ok(config) => config,
        Err(error) => panic!("Missing environment variables: {:#?}", error),
    };

    headers.insert("X-API-KEY", config.brewers_friend_api_key.parse().unwrap());

    let client = reqwest::Client::builder()
        .default_headers(headers)
        .build()?;

    Ok(client)
}

#[async_trait]
impl BFDataManager for BFApiDataManager {
    async fn sessions() -> Result<Vec<BFBrewSession>> {
        println!("Fetching sessions...");
        let client = create_client()?;

        let brew_sessions_url = format!("{}/v1/brewsessions", get_base_url());

        let res = client.get(&brew_sessions_url).send().await?;

        let body = res.text().await?;

        let response: BFBrewSessionsResponse = match serde_json::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        println!("Fetched [{:#?}] sessions", response.brewsessions.len());

        Ok(response.brewsessions)
    }

    async fn session(id: &str) -> Result<BFBrewSession> {
        println!("Fetching session with id [{}]...", id);
        let client = create_client()?;

        let brew_session_url = format!("{}/v1/brewsessions/{}", get_base_url(), id);

        let res = client.get(&brew_session_url).send().await?;

        let body = res.text().await?;

        let response: BFBrewSessionsResponse = match serde_json::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        let session = match response.brewsessions.as_slice() {
            [] => Err(Error::ApiResponseValidationError {
                message: String::from("No session with ID found"),
            }),
            [session] => Ok(session.to_owned()),
            [_args @ ..] => Err(Error::ApiResponseValidationError {
                message: String::from("Multiple sessions for ID found. This should not happen"),
            }),
        };
        println!("Fetched session with id [{}]...", id);

        session
    }

    async fn recipe(recipe_id: &str) -> Result<BeerXml> {
        println!("Fetching recipe with id [{}]...", recipe_id);
        let client = create_client()?;

        let brew_session_url = format!("{}/v1/recipes/{}.xml", get_base_url(), recipe_id);

        let res = client.get(&brew_session_url).send().await?;

        let body = res.text().await?;

        let response = match quick_xml::de::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        println!("Fetched recipe with id [{}]...", recipe_id);

        Ok(response)
    }
}
