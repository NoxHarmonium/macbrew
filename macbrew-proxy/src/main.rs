use error::{Error, Result};
use futures::sink::SinkExt;
use std::{env, str};
use tokio;
use tokio::stream::StreamExt;
use tokio_serial::{Serial, SerialPortSettings};
use tokio_util::codec::Framed;

mod codecs;
mod commands;
mod data;
mod error;
mod serializers;

#[cfg(unix)]
const DEFAULT_TTY: &str = "/dev/pts/1";
#[cfg(windows)]
const DEFAULT_TTY: &str = "COM1";

use codecs::line_codec::LineCodec;
use commands::command::{prepare_response, Command};
use data::brewers_friend::bf_api_data_manager::BFApiDataManager;

async fn handle_line(line: &str) -> Result<Vec<u8>> {
    let components = line.split(" ").collect::<Vec<_>>();
    let response = match components.as_slice() {
        [] => Err(Error::InvalidCommandInput {
            message: String::from(line),
        }),
        [rid, "LIST", "SESSION", args @ ..] => {
            commands::list_sessions::ListSessionsCommand::<BFApiDataManager>::handle(rid, args)
                .await
        }
        [rid, "GET", "SESSION", args @ ..] => {
            commands::get_sessions::GetSessionsCommand::<BFApiDataManager>::handle(rid, args).await
        }
        [rid, "GET", "RECIPE", args @ ..] => {
            commands::get_recipes::GetRecipesCommand::<BFApiDataManager>::handle(rid, args).await
        }
        [_, _args @ ..] => Err(Error::InvalidCommandInput {
            message: String::from(line),
        }),
    };
    match response {
        Ok(response) => Ok(response),
        Err(error) => prepare_response("0", false, &error),
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    let mut args = env::args();
    let tty_path = args.nth(1).unwrap_or_else(|| DEFAULT_TTY.into());

    let settings = SerialPortSettings::default();
    let mut port = Serial::from_path(tty_path, &settings).unwrap();

    #[cfg(unix)]
    port.set_exclusive(false)
        .expect("Unable to set serial port exclusive to false");

    let mut framed = Framed::new(port, LineCodec);

    while let Some(line_result) = framed.next().await {
        let line = line_result.expect("Failed to read line");
        let response = handle_line(line.trim()).await?;
        framed.send(response).await?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {

    use hexplay::HexViewBuilder;
    use mockito::mock;

    use super::*;

    fn format_binary_for_snap(binary: Vec<u8>) -> Vec<String> {
        let view = HexViewBuilder::new(&binary.as_slice())
            .row_width(16)
            .finish();
        // Split on new lines to make the snapshots easier to read
        view.to_string().split("\n").map(String::from).collect()
    }

    #[tokio::test]
    async fn test_get_recipe() {
        let recipe_xml = include_str!("../resources/sample_recipe.xml");
        let _m = mock("GET", "/v1/recipes/123456.xml")
            .with_status(200)
            .with_header("content-type", "application/xml")
            .with_header("x-api-key", "1234")
            .with_body(recipe_xml)
            .create();

        let result = handle_line("99 GET RECIPE 123456").await.unwrap();

        insta::assert_debug_snapshot!(format_binary_for_snap(result));
    }
}
