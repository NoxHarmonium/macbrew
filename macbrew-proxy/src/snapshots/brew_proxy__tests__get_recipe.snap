---
source: src/main.rs
expression: format_binary_for_snap(result)
---
[
    "00000000  02 00 00 00 39 39 01 0E 00 00 00 46 6F 72 72 65  | ☻...99☺♫...Forre |",
    "00000010  73 74 79 20 46 72 75 69 74 01 00 00 00 31 09 00  | sty Fruit☺...1○. |",
    "00000020  00 00 41 6C 6C 20 47 72 61 69 6E 0B 00 00 00 53  | ..All Grain♂...S |",
    "00000030  65 61 6E 20 44 61 77 73 6F 6E 00 02 00 00 00 31  | ean Dawson.☻...1 |",
    "00000040  33 04 00 00 00 32 38 2E 35 02 00 00 00 31 30 01  | 3♦...28.5☻...10☺ |",
    "00000050  02 00 00 00 35 35 0C 00 00 00 36 00 00 00 41 6D  | ☻...55♀...6...Am |",
    "00000060  61 72 69 6C 6C 6F 20 28 38 2E 36 20 41 41 29 0D  | arillo (8.6 AA)♪ |",
    "00000070  0A 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  | ◙                |",
    "00000080  20 20 20 20 20 20 20 20 20 2F 20 31 34 30 20 47  |          / 140 G |",
    "00000090  72 61 6D 73 01 00 00 00 31 03 00 00 00 38 2E 36  | rams☺...1♥...8.6 |",
    "000000A0  04 00 00 00 30 2E 30 31 04 00 00 00 4D 61 73 68  | ♦...0.01♦...Mash |",
    "000000B0  01 00 00 00 30 00 00 01 06 00 00 00 50 65 6C 6C  | ☺...0..☺♠...Pell |",
    "000000C0  65 74 00 00 00 00 00 00 00 00 01 04 00 00 00 4D  | et........☺♦...M |",
    "000000D0  61 73 68 00 00 00 3D 00 00 00 48 50 41 20 2D 20  | ash...=...HPA -  |",
    "000000E0  20 45 63 6C 69 70 73 65 20 28 31 36 2E 39 20 41  |  Eclipse (16.9 A |",
    "000000F0  41 29 0D 0A 20 20 20 20 20 20 20 20 20 20 20 20  | A)♪◙             |",
    "00000100  20 20 20 20 20 20 20 20 20 20 20 20 2F 20 31 30  |             / 10 |",
    "00000110  30 20 47 72 61 6D 73 01 00 00 00 31 04 00 00 00  | 0 Grams☺...1♦... |",
    "00000120  31 36 2E 39 05 00 00 00 30 2E 30 30 35 04 00 00  | 16.9♣...0.005♦.. |",
    "00000130  00 4D 61 73 68 01 00 00 00 30 00 00 01 06 00 00  | .Mash☺...0..☺♠.. |",
    "00000140  00 50 65 6C 6C 65 74 00 00 00 00 00 00 00 00 01  | .Pellet........☺ |",
    "00000150  04 00 00 00 4D 61 73 68 00 00 00 35 00 00 00 4C  | ♦...Mash...5...L |",
    "00000160  65 6D 6F 6E 64 72 6F 70 20 28 36 20 41 41 29 0D  | emondrop (6 AA)♪ |",
    "00000170  0A 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  | ◙                |",
    "00000180  20 20 20 20 20 20 20 20 20 2F 20 31 30 30 20 47  |          / 100 G |",
    "00000190  72 61 6D 73 01 00 00 00 31 01 00 00 00 36 04 00  | rams☺...1☺...6♦. |",
    "000001A0  00 00 30 2E 30 32 04 00 00 00 4D 61 73 68 01 00  | ..0.02♦...Mash☺. |",
    "000001B0  00 00 30 00 00 01 06 00 00 00 50 65 6C 6C 65 74  | ..0..☺♠...Pellet |",
    "000001C0  00 00 00 00 00 00 00 00 01 04 00 00 00 4D 61 73  | ........☺♦...Mas |",
    "000001D0  68 00 00 00 3F 00 00 00 4D 61 6E 64 61 72 69 6E  | h...?...Mandarin |",
    "000001E0  61 20 42 61 76 61 72 69 61 20 28 38 2E 35 20 41  | a Bavaria (8.5 A |",
    "000001F0  41 29 0D 0A 20 20 20 20 20 20 20 20 20 20 20 20  | A)♪◙             |",
    "00000200  20 20 20 20 20 20 20 20 20 20 20 20 2F 20 31 30  |             / 10 |",
    "00000210  30 20 47 72 61 6D 73 01 00 00 00 31 03 00 00 00  | 0 Grams☺...1♥... |",
    "00000220  38 2E 35 05 00 00 00 30 2E 30 30 35 04 00 00 00  | 8.5♣...0.005♦... |",
    "00000230  4D 61 73 68 01 00 00 00 30 00 00 01 06 00 00 00  | Mash☺...0..☺♠... |",
    "00000240  50 65 6C 6C 65 74 00 00 00 00 00 00 00 00 01 04  | Pellet........☺♦ |",
    "00000250  00 00 00 4D 61 73 68 00 00 00 11 00 00 00 41 6D  | ...Mash...◄...Am |",
    "00000260  61 72 69 6C 6C 6F 20 28 38 2E 36 20 41 41 29 01  | arillo (8.6 AA)☺ |",
    "00000270  00 00 00 31 03 00 00 00 38 2E 36 04 00 00 00 30  | ...1♥...8.6♦...0 |",
    "00000280  2E 30 31 05 00 00 00 41 72 6F 6D 61 02 00 00 00  | .01♣...Aroma☻... |",
    "00000290  31 35 00 00 01 06 00 00 00 50 65 6C 6C 65 74 00  | 15..☺♠...Pellet. |",
    "000002A0  00 00 00 00 00 00 00 01 09 00 00 00 57 68 69 72  | .......☺○...Whir |",
    "000002B0  6C 70 6F 6F 6C 01 02 00 00 00 37 35 01 00 00 00  | lpool☺☻...75☺... |",
    "000002C0  00 01 02 00 00 00 37 35 3D 00 00 00 48 50 41 20  | .☺☻...75=...HPA  |",
    "000002D0  2D 20 20 45 63 6C 69 70 73 65 20 28 31 36 2E 39  | -  Eclipse (16.9 |",
    "000002E0  20 41 41 29 0D 0A 20 20 20 20 20 20 20 20 20 20  |  AA)♪◙           |",
    "000002F0  20 20 20 20 20 20 20 20 20 20 20 20 20 20 2F 20  |               /  |",
    "00000300  31 30 30 20 47 72 61 6D 73 01 00 00 00 31 04 00  | 100 Grams☺...1♦. |",
    "00000310  00 00 31 36 2E 39 04 00 00 00 30 2E 30 31 05 00  | ..16.9♦...0.01♣. |",
    "00000320  00 00 41 72 6F 6D 61 02 00 00 00 31 35 00 00 01  | ..Aroma☻...15..☺ |",
    "00000330  06 00 00 00 50 65 6C 6C 65 74 00 00 00 00 00 00  | ♠...Pellet...... |",
    "00000340  00 00 01 09 00 00 00 57 68 69 72 6C 70 6F 6F 6C  | ..☺○...Whirlpool |",
    "00000350  01 02 00 00 00 37 35 01 00 00 00 00 01 02 00 00  | ☺☻...75☺....☺☻.. |",
    "00000360  00 37 35 10 00 00 00 4C 65 6D 6F 6E 64 72 6F 70  | .75►...Lemondrop |",
    "00000370  20 28 36 20 41 41 29 01 00 00 00 31 01 00 00 00  |  (6 AA)☺...1☺... |",
    "00000380  36 04 00 00 00 30 2E 30 31 05 00 00 00 41 72 6F  | 6♦...0.01♣...Aro |",
    "00000390  6D 61 02 00 00 00 31 35 00 00 01 06 00 00 00 50  | ma☻...15..☺♠...P |",
    "000003A0  65 6C 6C 65 74 00 00 00 00 00 00 00 00 01 09 00  | ellet........☺○. |",
    "000003B0  00 00 57 68 69 72 6C 70 6F 6F 6C 01 02 00 00 00  | ..Whirlpool☺☻... |",
    "000003C0  37 35 01 00 00 00 00 01 02 00 00 00 37 35 3F 00  | 75☺....☺☻...75?. |",
    "000003D0  00 00 4D 61 6E 64 61 72 69 6E 61 20 42 61 76 61  | ..Mandarina Bava |",
    "000003E0  72 69 61 20 28 38 2E 35 20 41 41 29 0D 0A 20 20  | ria (8.5 AA)♪◙   |",
    "000003F0  20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  |                  |",
    "00000400  20 20 20 20 20 20 2F 20 31 30 30 20 47 72 61 6D  |       / 100 Gram |",
    "00000410  73 01 00 00 00 31 03 00 00 00 38 2E 35 04 00 00  | s☺...1♥...8.5♦.. |",
    "00000420  00 30 2E 30 31 05 00 00 00 41 72 6F 6D 61 02 00  | .0.01♣...Aroma☻. |",
    "00000430  00 00 31 35 00 00 01 06 00 00 00 50 65 6C 6C 65  | ..15..☺♠...Pelle |",
    "00000440  74 00 00 00 00 00 00 00 00 01 09 00 00 00 57 68  | t........☺○...Wh |",
    "00000450  69 72 6C 70 6F 6F 6C 01 02 00 00 00 37 35 01 00  | irlpool☺☻...75☺. |",
    "00000460  00 00 00 01 02 00 00 00 37 35 36 00 00 00 41 6D  | ...☺☻...756...Am |",
    "00000470  61 72 69 6C 6C 6F 20 28 38 2E 36 20 41 41 29 0D  | arillo (8.6 AA)♪ |",
    "00000480  0A 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  | ◙                |",
    "00000490  20 20 20 20 20 20 20 20 20 2F 20 32 38 30 20 47  |          / 280 G |",
    "000004A0  72 61 6D 73 01 00 00 00 31 03 00 00 00 38 2E 36  | rams☺...1♥...8.6 |",
    "000004B0  04 00 00 00 30 2E 30 32 07 00 00 00 44 72 79 20  | ♦...0.02•...Dry  |",
    "000004C0  48 6F 70 04 00 00 00 35 37 36 30 00 00 01 06 00  | Hop♦...5760..☺♠. |",
    "000004D0  00 00 50 65 6C 6C 65 74 00 00 00 00 00 00 00 00  | ..Pellet........ |",
    "000004E0  01 16 00 00 00 44 72 79 20 48 6F 70 20 28 48 69  | ☺▬...Dry Hop (Hi |",
    "000004F0  67 68 20 4B 72 61 75 73 65 6E 29 00 00 00 3D 00  | gh Krausen)...=. |",
    "00000500  00 00 48 50 41 20 2D 20 20 45 63 6C 69 70 73 65  | ..HPA -  Eclipse |",
    "00000510  20 28 31 36 2E 39 20 41 41 29 0D 0A 20 20 20 20  |  (16.9 AA)♪◙     |",
    "00000520  20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  |                  |",
    "00000530  20 20 20 20 2F 20 31 30 30 20 47 72 61 6D 73 01  |     / 100 Grams☺ |",
    "00000540  00 00 00 31 04 00 00 00 31 36 2E 39 04 00 00 00  | ...1♦...16.9♦... |",
    "00000550  30 2E 30 31 07 00 00 00 44 72 79 20 48 6F 70 04  | 0.01•...Dry Hop♦ |",
    "00000560  00 00 00 35 37 36 30 00 00 01 06 00 00 00 50 65  | ...5760..☺♠...Pe |",
    "00000570  6C 6C 65 74 00 00 00 00 00 00 00 00 01 16 00 00  | llet........☺▬.. |",
    "00000580  00 44 72 79 20 48 6F 70 20 28 48 69 67 68 20 4B  | .Dry Hop (High K |",
    "00000590  72 61 75 73 65 6E 29 00 00 00 10 00 00 00 4C 65  | rausen)...►...Le |",
    "000005A0  6D 6F 6E 64 72 6F 70 20 28 36 20 41 41 29 01 00  | mondrop (6 AA)☺. |",
    "000005B0  00 00 31 01 00 00 00 36 04 00 00 00 30 2E 30 34  | ..1☺...6♦...0.04 |",
    "000005C0  07 00 00 00 44 72 79 20 48 6F 70 04 00 00 00 35  | •...Dry Hop♦...5 |",
    "000005D0  37 36 30 00 00 01 06 00 00 00 50 65 6C 6C 65 74  | 760..☺♠...Pellet |",
    "000005E0  00 00 00 00 00 00 00 00 01 16 00 00 00 44 72 79  | ........☺▬...Dry |",
    "000005F0  20 48 6F 70 20 28 48 69 67 68 20 4B 72 61 75 73  |  Hop (High Kraus |",
    "00000600  65 6E 29 00 00 00 3F 00 00 00 4D 61 6E 64 61 72  | en)...?...Mandar |",
    "00000610  69 6E 61 20 42 61 76 61 72 69 61 20 28 38 2E 35  | ina Bavaria (8.5 |",
    "00000620  20 41 41 29 0D 0A 20 20 20 20 20 20 20 20 20 20  |  AA)♪◙           |",
    "00000630  20 20 20 20 20 20 20 20 20 20 20 20 20 20 2F 20  |               /  |",
    "00000640  31 30 30 20 47 72 61 6D 73 01 00 00 00 31 03 00  | 100 Grams☺...1♥. |",
    "00000650  00 00 38 2E 35 04 00 00 00 30 2E 30 31 07 00 00  | ..8.5♦...0.01•.. |",
    "00000660  00 44 72 79 20 48 6F 70 04 00 00 00 35 37 36 30  | .Dry Hop♦...5760 |",
    "00000670  00 00 01 06 00 00 00 50 65 6C 6C 65 74 00 00 00  | ..☺♠...Pellet... |",
    "00000680  00 00 00 00 00 01 16 00 00 00 44 72 79 20 48 6F  | .....☺▬...Dry Ho |",
    "00000690  70 20 28 48 69 67 68 20 4B 72 61 75 73 65 6E 29  | p (High Krausen) |",
    "000006A0  00 00 00 03 00 00 00 10 00 00 00 4D 61 72 69 73  | ...♥...►...Maris |",
    "000006B0  20 4F 74 74 65 72 20 70 61 6C 65 01 00 00 00 31  |  Otter pale☺...1 |",
    "000006C0  05 00 00 00 47 72 61 69 6E 01 00 00 00 34 05 00  | ♣...Grain☺...4♣. |",
    "000006D0  00 00 38 32 2E 36 31 03 00 00 00 32 2E 34 01 05  | ..82.61♥...2.4☺♣ |",
    "000006E0  00 00 00 66 61 6C 73 65 01 0E 00 00 00 55 6E 69  | ...false☺♫...Uni |",
    "000006F0  74 65 64 20 4B 69 6E 67 64 6F 6D 00 00 00 00 01  | ted Kingdom....☺ |",
    "00000700  02 00 00 00 36 32 00 00 00 00 0B 00 00 00 42 45  | ☻...62....♂...BE |",
    "00000710  53 54 20 4D 75 6E 69 63 68 01 00 00 00 31 05 00  | ST Munich☺...1♣. |",
    "00000720  00 00 47 72 61 69 6E 04 00 00 00 30 2E 32 35 05  | ..Grain♦...0.25♣ |",
    "00000730  00 00 00 38 30 2E 34 33 03 00 00 00 36 2E 33 01  | ...80.43♥...6.3☺ |",
    "00000740  05 00 00 00 66 61 6C 73 65 01 07 00 00 00 47 65  | ♣...false☺•...Ge |",
    "00000750  72 6D 61 6E 79 00 00 00 00 01 02 00 00 00 37 30  | rmany....☺☻...70 |",
    "00000760  00 00 00 00 14 00 00 00 4C 61 63 74 6F 73 65 20  | ....¶...Lactose  |",
    "00000770  28 4D 69 6C 6B 20 53 75 67 61 72 29 01 00 00 00  | (Milk Sugar)☺... |",
    "00000780  31 05 00 00 00 53 75 67 61 72 03 00 00 00 30 2E  | 1♣...Sugar♥...0. |",
    "00000790  32 05 00 00 00 38 39 2E 31 33 01 00 00 00 31 01  | 2♣...89.13☺...1☺ |",
    "000007A0  04 00 00 00 74 72 75 65 00 00 00 00 00 01 01 00  | ♦...true.....☺☺. |",
    "000007B0  00 00 30 00 00 00 00 01 00 00 00 1E 00 00 00 4C  | ..0....☺...▲...L |",
    "000007C0  41 4C 42 52 45 57 C2 AE 20 56 4F 53 53 20 4B 56  | ALBREW┬« VOSS KV |",
    "000007D0  45 49 4B 20 41 4C 45 20 59 45 41 53 54 01 00 00  | EIK ALE YEAST☺.. |",
    "000007E0  00 31 04 00 00 00 41 6C 65 73 03 00 00 00 44 72  | .1♦...Ales♥...Dr |",
    "000007F0  79 04 00 00 00 30 2E 31 31 01 01 01 09 00 00 00  | y♦...0.11☺☺☺○... |",
    "00000800  4C 61 6C 6C 65 6D 61 6E 64 01 04 00 00 00 56 4F  | Lallemand☺♦...VO |",
    "00000810  53 53 01 02 00 00 00 32 35 01 02 00 00 00 34 30  | SS☺☻...25☺☻...40 |",
    "00000820  00 01 02 00 00 00 38 30 00 00 00 00 00 01 26 02  | .☺☻...80.....☺&☻ |",
    "00000830  00 00 53 65 63 6F 6E 64 20 69 74 65 72 61 74 69  | ..Second iterati |",
    "00000840  6F 6E 20 6F 66 20 74 72 79 69 6E 67 20 74 6F 20  | on of trying to  |",
    "00000850  6D 61 6B 65 20 61 20 76 65 72 79 20 63 69 74 72  | make a very citr |",
    "00000860  75 73 79 20 70 61 6C 65 20 61 6C 65 20 28 68 65  | usy pale ale (he |",
    "00000870  6E 63 65 20 74 68 65 20 70 75 6E 20 6F 6E 20 74  | nce the pun on t |",
    "00000880  68 65 20 46 72 6F 73 74 79 20 46 72 75 69 74 29  | he Frosty Fruit) |",
    "00000890  20 0D 0A 4C 61 73 74 20 74 69 6D 65 20 69 74 20  |  ♪◙Last time it  |",
    "000008A0  74 75 72 6E 65 64 20 6F 75 74 20 77 65 6C 6C 2C  | turned out well, |",
    "000008B0  20 61 6C 74 68 6F 75 67 68 20 69 74 20 73 65 65  |  although it see |",
    "000008C0  6D 65 64 20 74 6F 20 62 65 20 68 6F 70 70 79 20  | med to be hoppy  |",
    "000008D0  65 6E 6F 75 67 68 20 61 66 74 65 72 20 74 68 65  | enough after the |",
    "000008E0  20 66 69 72 73 74 20 64 72 79 20 68 6F 70 20 63  |  first dry hop c |",
    "000008F0  68 61 72 67 65 20 73 6F 20 49 27 6D 20 6D 6F 76  | harge so I\'m mov |",
    "00000900  69 6E 67 20 74 68 65 20 73 65 63 6F 6E 64 20 64  | ing the second d |",
    "00000910  72 79 20 68 6F 70 20 74 6F 20 61 20 6D 61 73 68  | ry hop to a mash |",
    "00000920  20 68 6F 70 20 28 74 6F 20 68 6F 70 65 66 75 6C  |  hop (to hopeful |",
    "00000930  6C 79 20 68 65 6C 70 20 77 69 74 68 20 63 6F 6C  | ly help with col |",
    "00000940  64 20 73 69 64 65 20 6F 78 69 64 61 74 69 6F 6E  | d side oxidation |",
    "00000950  29 20 61 6E 64 20 6F 6E 6C 79 20 64 6F 69 6E 67  | ) and only doing |",
    "00000960  20 68 69 67 68 20 6B 72 61 75 73 65 6E 20 64 72  |  high krausen dr |",
    "00000970  79 20 68 6F 70 2E 20 46 69 72 73 74 20 61 74 74  | y hop. First att |",
    "00000980  65 6D 70 74 20 77 69 74 68 20 6D 61 73 68 20 68  | empt with mash h |",
    "00000990  6F 70 70 69 6E 67 21 20 49 27 6C 6C 20 73 65 65  | opping! I\'ll see |",
    "000009A0  20 68 6F 77 20 69 74 20 67 6F 65 73 2E 0D 0A 0D  |  how it goes.♪◙♪ |",
    "000009B0  0A 41 6C 73 6F 20 74 72 79 69 6E 67 20 73 6F 6D  | ◙Also trying som |",
    "000009C0  65 20 4D 61 6E 64 61 72 69 6E 61 20 42 72 61 76  | e Mandarina Brav |",
    "000009D0  61 72 69 61 20 74 68 69 73 20 74 69 6D 65 20 61  | aria this time a |",
    "000009E0  6E 64 20 45 63 6C 69 70 73 65 20 28 48 50 41 2D  | nd Eclipse (HPA- |",
    "000009F0  30 31 36 29 20 77 68 69 63 68 20 61 70 70 61 72  | 016) which appar |",
    "00000A00  65 6E 74 6C 79 20 68 61 76 65 20 4D 61 6E 64 61  | ently have Manda |",
    "00000A10  72 69 6E 2F 54 61 6E 67 65 72 69 6E 65 20 66 6C  | rin/Tangerine fl |",
    "00000A20  61 76 6F 75 72 73 20 77 68 69 63 68 20 73 68 6F  | avours which sho |",
    "00000A30  75 6C 64 20 63 6F 6D 70 6C 65 6D 65 6E 74 20 74  | uld complement t |",
    "00000A40  68 65 20 6C 65 6D 6F 6E 2F 6F 72 61 6E 67 65 20  | he lemon/orange  |",
    "00000A50  66 6C 61 76 6F 75 72 73 00 00 01 04 00 00 00 31  | flavours..☺♦...1 |",
    "00000A60  2E 30 36 01 05 00 00 00 31 2E 30 31 36 00 00 01  | .06☺♣...1.016..☺ |",
    "00000A70  02 00 00 00 33 37 00 00 00 00 00 00 00 01 01 00  | ☻...37.......☺☺. |",
    "00000A80  00 00 30 00 01 00 00 00 00 00 00 00 01 04 00 00  | ..0.☺.......☺♦.. |",
    "00000A90  00 35 2E 38 34 01 05 00 00 00 31 32 2E 31 34 01  | .5.84☺♣...12.14☺ |",
    "00000AA0  07 00 00 00 54 69 6E 73 65 74 68 01 04 00 00 00  | •...Tinseth☺♦... |",
    "00000AB0  35 2E 37 37 01 07 00 00 00 31 2E 30 36 20 73 67  | 5.77☺•...1.06 sg |",
    "00000AC0  01 08 00 00 00 31 2E 30 31 36 20 73 67           | ☺◘...1.016 sg    |",
]
